### PR TITLE
LKMEX Governance

### PR DESCRIPTION
**Pull request by ccorcoveanu**
Changed only needed functionality so that instead of energy:
- user sends the voting power himself (before, energy was queried from the energy SC)
- additionally the user needs to pass a correct merkle proof for that VP
- contract verifies the proof against a stored root hash 
Note: When generating the merkle tree, all hashes need to be sorted so that the verification function is as simple as possible

### CODE REVIEW

File: energy-integration/governance-v2/meta/Cargo.toml [hunk 1]
```diff
@@ -11,4 +11,4 @@ authors = [ "you",]
 path = ".."
 
 [dependencies.elrond-wasm-debug]
-version = "=0.36.1"
+version = "0.36.1"
```
Comment (dorin-iancu) [inline energy-integration/governance-v2/meta/Cargo.toml:14]: Leave the `=0.36.1` please, this ensures we use that specific version.

File: energy-integration/governance-v2/src/lib.rs [hunk 2]
```diff
@@ -143,7 +139,7 @@ pub trait GovernanceV2:
 
     /// Vote on a proposal. The voting power depends on the user's energy.
     #[endpoint]
-    fn vote(&self, proposal_id: ProposalId, vote: VoteType) {
+    fn vote(&self, proposal_id: ProposalId, vote: VoteType, power: BigUint<Self::Api>, proof: ArrayVec<ManagedByteArray<HASH_LENGTH>, PROOF_LENGTH>) {
```
Comment (dorin-iancu) [inline energy-integration/governance-v2/src/lib.rs:142]: run `cargo fmt` to format the code

File: energy-integration/governance-v2/src/lib.rs [hunk 3]
```diff
@@ -341,4 +348,29 @@ pub trait GovernanceV2:
 
         self.proposal_votes(proposal_id).clear();
     }
+
+    fn verify_merkle_proof(&self, power: BigUint<Self::Api>, proof: ArrayVec<ManagedByteArray<HASH_LENGTH>, PROOF_LENGTH>, root_hash: ManagedByteArray<HASH_LENGTH>) -> bool {
```
Comment (dorin-iancu) [inline energy-integration/governance-v2/src/lib.rs:352]: Verification logic should be a separate module

File: energy-integration/governance-v2/src/lib.rs [hunk 4]
```diff
@@ -341,4 +348,29 @@ pub trait GovernanceV2:
...
+                hash = self.crypto().sha256(tst);
+            } else {
+                let mut tst = proof_item.as_managed_buffer().clone();
+                tst.append(hash.as_managed_buffer());
+
+                hash = self.crypto().sha256(tst);
+            }
+        }
+
+        return hash == root_hash;
```
Comment (dorin-iancu) [inline energy-integration/governance-v2/src/lib.rs:374]:
```suggestion
        hash == root_hash
```

File: energy-integration/governance-v2/meta/Cargo.toml (hunk 1 above)
Comment (ccorcoveanu) [reply energy-integration/governance-v2/meta/Cargo.toml:14]: done

File: energy-integration/governance-v2/src/lib.rs [hunk 5]
```diff
@@ -71,23 +71,18 @@ pub trait GovernanceV2:
     #[endpoint]
     fn propose(
         &self,
+        root_hash: ManagedByteArray<HASH_LENGTH>,
```
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply energy-integration/governance-v2/src/lib.rs:75]: added onlu_owner

**Pull request closed without merging**